server/src/main/java/org/opensearch/action/search/SearchQueryCategorizor.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryBuilderVisitor;
+import org.opensearch.index.query.QueryShapeVisitor;
+import org.opensearch.search.aggregations.AggregatorFactories;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.search.sort.SortBuilder;
+import org.opensearch.telemetry.metrics.MetricsRegistry;
+
+import java.util.List;
+
+/**
+ * Class to categorize the search queries based on the type and increment the relevant counters.
+ * Class also logs the query shape.
+ */
+public class SearchQueryCategorizor {
+
+    private static final Logger log = LogManager.getLogger(SearchQueryCategorizor.class);
+
+    public static SearchQueryCounters searchQueryCounters;
```

Comment:
Should this be static? A `public static` variable is essentially a global variable. In this case, it's overwritten on every construction. 

Will we have more than one `SearchQueryCategorizer`? Should this be a singleton owned by one of the `Service` classes? Maybe `SearchService`?

Reply:
We will have only one SearchQueryCategorizer which is initialized in TransportSearchAction and behind a feature flag